### PR TITLE
step over rep and repne prefixed instructions with dsui

### PR DESCRIPTION
**Pull request by kamou**

### CODE REVIEW

**Comment by radare** [comment]
So this patch works?

File: libr/core/cmd_debug.c [hunk 1]
```diff
@@ -790,7 +804,12 @@ static int step_until_inst(RCore *core, const char *instr, bool regex) {
 		if (r_debug_is_dead (core->dbg)) {
 			break;
 		}
-		r_debug_step (core->dbg, 1);
+		if (is_repeatalbe_inst (core->dbg)) {
+			r_debug_step_over (core->dbg, 1);
+		}
```
Comment (radare) [inline libr/core/cmd_debug.c]: Wrong indent

File: libr/core/cmd_debug.c [hunk 2]
```diff
@@ -790,7 +804,12 @@ static int step_until_inst(RCore *core, const char *instr, bool regex) {
 		if (r_debug_is_dead (core->dbg)) {
 			break;
 		}
-		r_debug_step (core->dbg, 1);
+		if (is_repeatalbe_inst (core->dbg)) {
```
Comment (radare) [inline libr/core/cmd_debug.c]: Typo

File: libr/core/cmd_debug.c [hunk 3]
```diff
@@ -771,6 +771,20 @@ static int step_until_esil(RCore *core, const char *esilstr) {
 	return true;
 }
 
+static bool is_repeatalbe_inst(RDebug *dbg) {
+	RAnalOp op;
```
Comment (radare) [inline libr/core/cmd_debug.c]: Return false if arch is not. X86

File: libr/core/cmd_debug.c [hunk 4]
```diff
@@ -771,6 +771,20 @@ static int step_until_esil(RCore *core, const char *esilstr) {
...
+	ut64  pc;
+	ut8 buf[32];
+	pc = r_debug_reg_get (dbg, dbg->reg->name[R_REG_NAME_PC]);
```
Comment (radare) [inline libr/core/cmd_debug.c]: Theres a function in core that gives u a ranalop by giving the address. Nonneed to do the ioread+anal by hand every time

Comment (kamou) [reply libr/core/cmd_debug.c]: done

File: libr/core/cmd_debug.c (hunk 3 above)
Comment (kamou) [reply libr/core/cmd_debug.c]: done

File: libr/core/cmd_debug.c (hunk 2 above)
Comment (kamou) [reply libr/core/cmd_debug.c]: done

File: libr/core/cmd_debug.c (hunk 1 above)
Comment (kamou) [reply libr/core/cmd_debug.c]: sorry, maybe I'm too tired, but can't see any indent issue.

**Comment by kamou** [comment]
> So this patch works?

yes

**Comment by kamou** [comment]
> This is making dsu faster when stepping over rep instructions iirc, but slower in all the other cases. This should be x86 specific. Also , assign and define the vars in the same line if possible

I moved RAnalOp allocatin after checking of arch for the sake of speed.

File: libr/core/cmd_debug.c [hunk 5]
```diff
@@ -771,6 +771,25 @@ static int step_until_esil(RCore *core, const char *esilstr) {
 	return true;
 }
 
+static bool is_repeatable_inst(RCore *core, ut64 addr) {
+	bool ret = false;
+
+	if (strcmp (r_config_get (core->config, "asm.arch"), "x86")) {
```
Comment (radare) [inline libr/core/cmd_debug.c]: use r_str_startswith() instead

File: libr/core/cmd_debug.c [hunk 6]
```diff
@@ -771,6 +771,25 @@ static int step_until_esil(RCore *core, const char *esilstr) {
...
+		goto end;
```
Comment (radare) [inline libr/core/cmd_debug.c]: avoid using gotos

File: libr/core/cmd_debug.c (hunk 1 above)
Comment (radare) [reply libr/core/cmd_debug.c]: the "else" must be in the lline above "} else {"

**Comment by radare** [comment]
Please rebase

**Comment by radare** [comment]
ping

**Comment by radare** [comment]
Screenshot 2019-07-22 at 22 59 42

File: libr/core/cmd_debug.c [hunk 7]
```diff
@@ -771,6 +771,25 @@ static int step_until_esil(RCore *core, const char *esilstr) {
...
 }
 
+static bool is_repeatable_inst(RCore *core, ut64 addr) {
+	bool ret = false;
+
+	if (!r_str_startswith (r_config_get (core->config, "asm.arch"), "x86")) {
+		return false;
+	}
+
+	RAnalOp *op = r_core_op_anal (core, addr);
```
Comment (radare) [inline libr/core/cmd_debug.c]: memleaks. you have to free this thing after using it

File: libr/core/cmd_debug.c [hunk 8]
```diff
@@ -771,6 +771,25 @@ static int step_until_esil(RCore *core, const char *esilstr) {
...
+	if (!op) {
+		eprintf ("Cannot analyze opcode at 0x%08" PFMT64x "\n", addr);
+		return false;
+	}
+
+	ret = (op->prefix & R_ANAL_OP_PREFIX_REP) || (op->prefix & R_ANAL_OP_PREFIX_REPNE);
+
+end:
```
Comment (radare) [inline libr/core/cmd_debug.c]: unnecessary label

**Comment by radare** [comment]
rebase again because i merged some new tests and will be red otherwise, but lgtm despite the comments

**Comment by radare** [comment]
Also, i think this thing should also obey cfg.r2wars config variable, otherwise the game will probably not work as expected, we should add tests for this cc @sanguinawer

**Comment by kamou** [comment]
> Also, i think this thing should also obey cfg.r2wars config variable, otherwise the game will probably not work as expected, we should add tests for this cc @sanguinawer

I'm not sure to understand the request. should I avoid stepping over when cfg.r2wars is set ?

**Comment by radare** [comment]
Yes

**Comment by radare** [comment]
But r2wars doesnt use dsui... so meh

**Pull request merged**